### PR TITLE
Fix string constructor for ShellNavigationState

### LINKED ISSUE #13422: [Bug] Location property no longer set via ShellNavigationState's Uri ctor
### Description

Constructing a `ShellNavigationState` instance from a `Uri` would set its `Location` property in Xamarin Forms v4.8 (_via its internal `FullLocation` property_). This no longer occurs in v5.0, so the value is `null`.

### Steps to Reproduce

1. Create a new Xamarin Forms (_v5_) Mobile App Shell project.
2. In App's `ctor`, create two `ShellNavigationState` instances: one from a `Uri` and the other from its string value.
3. Observe that `Location` is set for the second instance but not the first.
4. Revert to Xamarin Forms v4.8 and observe that both `Location` values are correct.

### Expected Behavior

`Location` should reflect the `Uri` given.

### Actual Behavior

`Location` is `null`.

### Basic Information

- Version with issue: 5.0.0.1874
- Last known good version: 4.8.0.1821
- Platform Target Frameworks: 
  - Android: 10.0  
- NuGet Packages: Defaults

### Environment

<details><summary>Show/Hide Visual Studio info</summary>

```
Microsoft Visual Studio Community 2019
Version 16.8.4
VisualStudio.16.Release/16.8.4+30907.101
Microsoft .NET Framework
Version 4.8.04084

Installed Version: Community

Visual C++ 2019   00435-60000-00000-AA479
Microsoft Visual C++ 2019

ASP.NET and Web Tools 2019   16.8.557.25636
ASP.NET and Web Tools 2019

ASP.NET Core Razor Language Services   16.1.0.2052803+84e121f1403378489b842e1797df2f3f5a49ac3c
Provides languages services for ASP.NET Core Razor.

Automatic Versions   2.0.48
Provides tools for managing assembly versions.

Azure App Service Tools v3.0.0   16.8.557.25636
Azure App Service Tools v3.0.0

C# Tools   3.8.0-5.20604.10+9ed4b774d20940880de8df1ca8b07508aa01c8cd
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.2.6 (master@34d6af2)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   2.1.113+g422d40002e.RR
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

Mono Debugging for Visual Studio   16.8.43 (00471f8)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   5.8.1
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

SQL Server Data Tools   16.0.62012.31170
Microsoft SQL Server Data Tools

TypeScript Tools   16.0.21016.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.8.0-5.20604.10+9ed4b774d20940880de8df1ca8b07508aa01c8cd
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools   16.8.0-beta.20507.4+da6be68280c89131cdba2045525b80890401defd
Microsoft Visual F# Tools

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Container Tools Extensions   1.0
View, manage, and diagnose containers within Visual Studio.

Visual Studio Spell Check Everywhere   VSSpellCheckEverywhere
An extension that enables spell checking within any Visual Studio file editor or tool window that uses WPF text boxes.
https://GitHub.com/EWSoftware/VSSpellChecker

Visual Studio Spell Checker   VSSpellChecker
An editor extension that checks the spelling of comments, strings, and plain text as you type or interactively with tool windows.
https://GitHub.com/EWSoftware/VSSpellChecker

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

Visual Studio Tools for Unity   4.8.2.0
Visual Studio Tools for Unity

VisualStudio.DeviceLog   1.0
Information about my package

VisualStudio.Foo   1.0
Information about my package

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   16.8.000.261 (d16-8@bb99248)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   16.8.0.507 (remotes/origin/d16-8@e87b24884)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   16.8.112 (86385a3)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   11.1.0.26 (d16-8/a36ce73)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: 5e9cb6d
    Java.Interop: xamarin/java.interop/d16-8@79d9533
    ProGuard: Guardsquare/proguard/proguard6.2.2@ebe9000
    SQLite: xamarin/sqlite/3.32.1@1a3276b
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-8@2fb1cbc


Xamarin.iOS and Xamarin.Mac SDK   14.8.0.3 (c51fabee8)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.
```

</details>

### Reproduction Link

```
public App()
{
    InitializeComponent();

    MainPage = new MainPage();

    var uri = new Uri("//test", UriKind.Relative);
    var uriState = new ShellNavigationState(uri);
    var strState = new ShellNavigationState(uri.ToString());
    var uriLoc = uriState.Location;
    var strLoc = strState.Location;
}
```

### Workaround

Pass the `Uri`'s string version (_i.e. `ToString`_) into `ShellNavigationState`'s `ctor`.

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###

Fix string constructor for ShellNavigationState so it sets both Location and FullLocation

### Issues Resolved ### 

- fixes #13422 

### Platforms Affected ### 
- Core/XAML (all platforms)

### Testing Procedure ###
- unit tests included

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Core.UnitTests/ShellNavigatingTests.cs [hunk 1]
```diff
@@ -637,6 +643,296 @@ public async Task InitialNavigatingArgs()
 				null, "//item");
 		}
 
+
```
Comment (PureWeen) [inline Xamarin.Forms.Core.UnitTests/ShellNavigatingTests.cs:646]: All of this is just moving code to the correct file. No changes happen here

File: Xamarin.Forms.Core.UnitTests/ShellNavigationStateTests.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+using NUnit.Framework;
+using Xamarin.Forms.Internals;
+
+namespace Xamarin.Forms.Core.UnitTests
+{
+	[TestFixture]
+	public class ShellNavigationStateTests : ShellTestBase
+	{
+		[Test]
+		public void LocationInitializedWithUri()
```
Comment (PureWeen) [inline Xamarin.Forms.Core.UnitTests/ShellNavigationStateTests.cs:16]: These Tests are new

File: Xamarin.Forms.Core.UnitTests/ShellUriHandlerTests.cs [hunk 3]
```diff
@@ -16,201 +16,6 @@ public override void TearDown()
 			Routing.Clear();
 		}
 
-		[Test]
-		public async Task RouteWithGlobalPageRoute()
```
Comment (PureWeen) [inline Xamarin.Forms.Core.UnitTests/ShellUriHandlerTests.cs:20]: All of this is just moving code to the correct file. No changes happen here

File: Xamarin.Forms.Core/Shell/ShellNavigationState.cs [hunk 4]
```diff
@@ -45,7 +45,12 @@ internal ShellNavigationState(string location, bool trimForUser)
 				Location = FullLocation;
 		}
 
-		public ShellNavigationState(Uri location) => FullLocation = location;
+		public ShellNavigationState(Uri location)
+		{
+			FullLocation = location;
+			Location = TrimDownImplicitAndDefaultPaths(FullLocation);
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Shell/ShellNavigationState.cs:51]: This is the primary change from this PR

File: Xamarin.Forms.Core/Shell/Shell.cs [hunk 5]
```diff
@@ -852,7 +852,7 @@ protected override bool OnBackButtonPressed()
 				return true;
 			}
 
-			var args = new ShellNavigatingEventArgs(this.CurrentState, "", ShellNavigationSource.Pop, true);
+			var args = new ShellNavigatingEventArgs(this.CurrentState, new ShellNavigationState(""), ShellNavigationSource.Pop, true);
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Shell/Shell.cs:855]: This change is just so we're avoiding using the implicit operator

File: Xamarin.Forms.Core/Shell/ShellNavigationManager.cs [hunk 6]
```diff
@@ -36,7 +36,7 @@ public async Task GoToAsync(ShellNavigationParameters shellNavigationParameters)
 			if (shellNavigationParameters.PagePushing != null)
 				Routing.RegisterImplicitPageRoute(shellNavigationParameters.PagePushing);
 
-			ShellNavigationState state = shellNavigationParameters.TargetState ?? Routing.GetRoute(shellNavigationParameters.PagePushing);
+			var state = shellNavigationParameters.TargetState ?? new ShellNavigationState(Routing.GetRoute(shellNavigationParameters.PagePushing), false);
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Shell/ShellNavigationManager.cs:39]: This change is just so we're avoiding using the implicit operator

File: Xamarin.Forms.Core/Shell/ShellSection.cs [hunk 7]
```diff
@@ -1063,7 +1063,7 @@ protected override async Task<Page> OnPopAsync(bool animated)
 				var navigationParameters = new ShellNavigationParameters()
 				{
 					Animated = animated,
-					TargetState = ".."
+					TargetState = new ShellNavigationState("..")
```
Comment (PureWeen) [inline Xamarin.Forms.Core/Shell/ShellSection.cs:1066]: This change is just so we're avoiding using the implicit operator

**Comment by PureWeen** [comment]
Failing tests not related

**Review by hartez — APPROVED** [review]

**Review by rmarinho — APPROVED** [review]

**Pull request merged**